engine/docker/docker.go

```diff
@@ -157,10 +157,31 @@ func (e *Docker) Destroy(ctx context.Context, pipelineConfig *spec.PipelineConfi
 	containers := e.containers
 	e.mu.Unlock()
 
+	timeout := time.Second * 30 // similar to k8s default timeout in pods
```

Comment:
I would be cautious with introducing this for all flows. This lite-engine is currently used in our hosted builds and the destroy flow there does not need to be graceful. If we are changing that behaviour, could we add a flag in the pipeline config which could dictate whether it's a graceful destroy or a force destroy (and it could be force by default and graceful for IACM?). That would make this change a lot safer and it would get some soak time IMO.

Reply:
So I do agree with you but it means that we probably need to expand the API and that is where @tphoney or you can help. We could 
- New method, instead of Destroy, Stop. Then we have the logic there and people could use destroy if they want to, or stop. The problem of this is:
- Because is a new API we probably need something to determine if we need to use Destroy or Stop. I guess that is a change in the runners?

or

- We could expand the Destroy. Maybe spec.PipelineConfig can have a new field, something around:
- - Destroy bool -> if true use destroy path if not, follow the stop path

Any suggestion are really appreciated and I am happy to change/implement them but I would need guidance of what needs to be done and changed

Reply:
IMO adding a flag in the container spec which gets propagated from the IACM step would be helpful here (this would involve a change in the runner/java code as well). The APIs and methods could remain the same, but if `force` is set to `false` for a container in the Destroy config, we would first try to do a `ContainerStop` with a timeout as opposed to a `ContainerKill`.
This would give some soak time for the change in IACM and not affect the rest of the flows that exist today. Let me know your thoughts.

Reply:
So, I like your idea but the problem that I see is that, the destroy step only have a parameter which is the pipelineConfig object. This seems to be a configuration object that is propagated from the runner to the lite-engine. The `ContainerRemoveOptions` seems to be a struct to control the removing of the containers, not the way to stop them.

With that said, i am experimenting now with something where, the InitialiseStep, which is the step  that contains the values for the `spec.PipelineConfig` would loop through the steps and if any of the steps is an apply or a destroy, then we set an envvar called `HARNESS_SOFT_STOP`. Then I could use that envvar which should be present in the destroy method to then choose if i want to perform a stop, or a kill. Then I was thinking to abstract the stop to another function so basically the code will be

if envar {
use stop method and don't finish until all the containers are stopped
} else {
kill as it was before
}

remove containers <- at this point if we executed the stop method all the containers should have been stopped.

By doing it with envvars i can keep the API methods the same although it feels a bit hacky to me.

If you know another way to do it I am happy to have a quick call and we can discuss it